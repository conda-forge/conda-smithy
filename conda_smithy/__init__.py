try:
    from conda_smithy._version import __version__
except ImportError:
    __version__ = "0.0.0"
