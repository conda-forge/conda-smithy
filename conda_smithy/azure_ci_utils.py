import os
import typing
import warnings

from msrest.authentication import Authentication, BasicAuthentication
from vsts.build.v4_1.build_client import BuildClient
from vsts.build.v4_1.models import (
    BuildDefinition,
    BuildDefinitionReference,
    SourceRepositories,
    SourceRepository,
)
from vsts.service_endpoint.v4_1.models import ServiceEndpoint
from vsts.service_endpoint.v4_1.service_endpoint_client import (
    ServiceEndpointClient,
)
from vsts.task_agent.v4_0.models import TaskAgentQueue
from vsts.task_agent.v4_0.task_agent_client import TaskAgentClient
from vsts.vss_connection import VssConnection


class AzureConfig:

    _default_org = "conda-forge"
    _default_project_name = "feedstock-builds"

    def __init__(
        self, org_or_user=None, project_name=None, team_instance=None
    ):
        self.org_or_user = org_or_user or os.getenv(
            "AZURE_ORG_OR_USER", self._default_org
        )
        # This will only need to be changed if you need to point to a non-standard azure
        # instance.
        self.instance_base_url = team_instance or os.getenv(
            "AZURE_INSTANCE", f"https://dev.azure.com/{self.org_or_user}"
        )
        self.project_name = project_name or os.getenv(
            "AZURE_PROJECT_NAME", self._default_project_name
        )

        try:
            with open(
                os.path.expanduser("~/.conda-smithy/azure.token"), "r"
            ) as fh:
                self.token = fh.read().strip()
            if not self.token:
                raise ValueError()
        except (IOError, ValueError):
            print(
                "No azure token. Create a token and\n"
                "put it in ~/.conda-smithy/azure.token"
            )
            self.token = None

        # By default for now don't report on the build information back to github
        self.azure_report_build_status = os.getenv(
            "AZURE_REPORT_BUILD_STATUS", "true"
        )

    @property
    def connection(self):
        connection = VssConnection(
            base_url=self.instance_base_url, creds=self.credentials
        )
        return connection

    @property
    def credentials(self):
        if self.token:
            return BasicAuthentication("", self.token)
        else:
            warnings.warn(
                "No token available.  No modifications will be possible!"
            )
            return Authentication()


default_config = AzureConfig()


def get_service_endpoint(config: AzureConfig = default_config):
    service_endpoint_client = ServiceEndpointClient(
        base_url=config.instance_base_url, creds=config.credentials
    )
    endpoints: typing.List[
        ServiceEndpoint
    ] = service_endpoint_client.get_service_endpoints(
        project=config.project_name, type="GitHub"
    )
    for service_endpoint in endpoints:
        if service_endpoint.name == config.org_or_user:
            return service_endpoint
    else:
        raise KeyError("Service endpoint not found")


def get_queues(
    config: AzureConfig = default_config,
) -> typing.List[TaskAgentQueue]:
    aclient = TaskAgentClient(config.instance_base_url, config.credentials)
    return aclient.get_agent_queues(config.project_name)


def get_default_queue(project_name):
    queues = get_queues(project_name)
    for q in queues:
        if q.name == "Default":
            return q
    else:
        raise ValueError("Default queue not found")


def get_repo_reference(config: AzureConfig, github_org, repo_name):
    service_endpoint = get_service_endpoint(config)
    bclient: BuildClient = config.connection.get_client(
        "vsts.build.v4_1.build_client.BuildClient"
    )
    repos: SourceRepositories = bclient.list_repositories(
        project=config.project_name,
        provider_name="github",
        repository=f"{github_org}/{repo_name}",
        service_endpoint_id=service_endpoint.id,
    )

    repo: SourceRepository = repos.repositories[0]
    return repo


def get_default_build_definition(
    github_org, repo_name, config: AzureConfig = default_config, **kwargs,
):
    from vsts.build.v4_1.models import (
        BuildDefinition,
        BuildRepository,
    )
    from vsts.task_agent.v4_0.task_agent_client import TaskAgentClient
    import inspect

    aclient = TaskAgentClient(config.instance_base_url, config.credentials)

    source_repo = get_repo_reference(config, github_org, repo_name)

    new_repo = BuildRepository(
        type="GitHub",
        url=source_repo.properties["cloneUrl"],
        **{
            k: v
            for k, v in source_repo.as_dict().items()
            if k in set(inspect.getfullargspec(BuildRepository).args) - {"url"}
        },
    )
    new_repo.name = source_repo.properties["fullName"]
    new_repo.properties["cleanOptions"] = "0"
    new_repo.properties["skipSyncSource"] = "false"
    new_repo.properties["gitLfsSupport"] = "false"
    new_repo.properties["checkoutNestedSubmodules"] = "false"
    new_repo.properties["labelSources"] = "0"
    new_repo.properties["fetchDepth"] = "0"
    new_repo.properties["labelSourcesFormat"] = "$(build.buildNumber)"
    new_repo.properties["reportBuildStatus"] = config.azure_report_build_status
    new_repo.clean = False

    queues = get_queues(config)
    default_queue = get_default_queue(config)
    service_endpoint = get_service_endpoint(config)

    build_definition = BuildDefinition(
        process={
            "type": 2,
            # These might be optional;
            "resources": {
                "queues": [{"id": q.id, "alias": q.name} for q in queues],
                "endpoints": [
                    {"id": service_endpoint.id, "alias": service_endpoint.name}
                ],
            },
            "yamlFilename": "/azure-pipelines.yml",
        },
        # queue works
        queue=default_queue,
        # now onto this
        repository=new_repo,
        name=repo_name,
        # configure trigger for our builds.
        triggers=[
            {
                "branchFilters": ["+*"],
                "forks": {"enabled": True, "allowSecrets": False},
                "pathFilters": [],
                "isCommentRequiredForPullRequest": False,
                "triggerType": "pullRequest",
            },
            {
                "branchFilters": ["+*"],
                "pathFilters": [],
                "batchChanges": False,
                "maxConcurrentBuildsPerBranch": 1,
                "pollingInterval": 0,
                "triggerType": "continuousIntegration",
            },
        ],
        variable_groups=aclient.get_variable_groups(
            project=config.project_name, group_name="anaconda-org"
        ),
        type="build",
        **kwargs,
    )

    return build_definition


def register_repo(github_org, repo_name, config: AzureConfig = default_config):
    from vsts.build.v4_1.models import BuildDefinitionReference

    bclient = build_client()
    build_definition = get_default_build_definition(
        github_org, repo_name, config=config,
    )

    # clean up existing builds for the same feedstock if present
    existing_definitions: typing.List[
        BuildDefinitionReference
    ] = bclient.get_definitions(project=config.project_name, name=repo_name)
    if existing_definitions:
        assert len(existing_definitions) == 1
        ed = existing_definitions[0]
        bclient.update_definition(
            definition=build_definition,
            definition_id=ed.id,
            project=ed.project.name,
        )
    else:
        bclient.create_definition(
            definition=build_definition, project=config.project_name
        )


def build_client(config: AzureConfig = default_config) -> BuildClient:
    return config.connection.get_client(
        "vsts.build.v4_1.build_client.BuildClient"
    )


def repo_registered(
    github_org: str, repo_name: str, config: AzureConfig = default_config
) -> bool:
    existing_definitions: typing.List[BuildDefinitionReference] = build_client(
        config
    ).get_definitions(project=config.project_name, name=repo_name)

    return bool(existing_definitions)


def enable_reporting(repo, config: AzureConfig = default_config) -> None:
    bclient = build_client(config)
    bdef_header = bclient.get_definitions(
        project=config.project_name, name=repo
    )[0]
    bdef = bclient.get_definition(bdef_header.id, bdef_header.project.name)
    bdef.repository.properties["reportBuildStatus"] = "true"
    bclient.update_definition(bdef, bdef.id, bdef.project.name)


def get_build_id(repo, config: AzureConfig = default_config) -> dict:
    """Get the necessary build information to persist in the config file to allow rendering
    of badges.
    This is needed by non-conda-forge use cases"""
    bclient = build_client(config)
    bdef_header = bclient.get_definitions(
        project=config.project_name, name=repo
    )[0]
    bdef: BuildDefinition = bclient.get_definition(
        bdef_header.id, bdef_header.project.name
    )

    return dict(
        user_or_org=config.org_or_user,
        project_name=config.project_name,
        build_id=bdef.id,
        project_id=bdef.project.id,
    )
