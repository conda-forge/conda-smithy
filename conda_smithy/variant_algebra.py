"""Variant algebras

This set of utilities are used to compose conda-build variants in a consistent way in
order to facilitate storing migration state within recipes rather than relying on
global state stored in something like `conda-forge-pinning`

The primary function to run here is ``variant_add`` that will add two variant configs
together and produce a desired outcome.

For full details on how this is supposed to work see CFEP-9

https://github.com/conda-forge/conda-forge-enhancement-proposals/pull/13

"""

import yaml
import toolz
from conda_build.utils import ensure_list
import conda_build.variants as variants
from conda.exports import VersionOrder
from conda_build.config import Config
from functools import partial


from typing import Any, Dict, List, Optional, Union


def parse_variant(
    variant_file_content: str, config: Optional[Config] = None
) -> Dict[
    str,
    Union[
        List[str],
        float,
        List[List[str]],
        Dict[str, Dict[str, str]],
        Dict[str, Dict[str, List[str]]],
    ],
]:
    """
    Parameters
    ----------
    variant_file_content : str
        The loaded vaiant contents.  This can include selectors etc.
    """
    if not config:
        from conda_build.config import Config

        config = Config()
    from conda_build.metadata import select_lines, ns_cfg

    contents = select_lines(
        variant_file_content, ns_cfg(config), variants_in_place=False
    )
    content = yaml.load(contents, Loader=yaml.loader.BaseLoader) or {}
    variants.trim_empty_keys(content)
    # TODO: Base this default on mtime or something
    content["migration_ts"] = float(content.get("migration_ts", -1.0))
    return content


def _version_order(
    v: Union[str, float], ordering: Optional[List[str]] = None
) -> Union[int, VersionOrder, float]:
    if ordering is not None:
        return ordering.index(v)
    else:
        try:
            return VersionOrder(v)
        except:
            return v


def variant_key_add(
    k: str,
    v_left: Union[List[str], List[float]],
    v_right: Union[List[str], List[float]],
    ordering: Optional[List[str]] = None,
) -> Union[List[str], List[float]]:
    """Version summation adder.

    This takes the higher version of the two things.
    """
    out_v = []
    common_length = min(len(v_left), len(v_right))
    for i in range(common_length):
        e_l, e_r = v_left[i], v_right[i]
        if _version_order(e_l, ordering) < _version_order(e_r, ordering):
            out_v.append(e_r)
        else:
            out_v.append(e_l)
    # Tail items
    for vs in (v_left, v_right):
        if len(vs) > common_length:
            out_v.extend(vs[common_length:])

    return out_v


def variant_key_replace(k, v_left, v_right):
    return v_right


def variant_key_set_merge(k, v_left, v_right, ordering=None):
    """Merges two sets in order"""
    out_v = set(v_left) & set(v_right)
    return sorted(out_v, key=partial(_version_order, ordering=ordering))


def variant_add(v1: dict, v2: dict) -> Dict[str, Any]:
    """Adds the two variants together.

    Present this assumes mostly flat dictionaries.

    TODO:
        - Add support for special sums like replace
        - Add delete_key support
    """
    left = set(v1.keys()).difference(v2.keys())
    right = set(v2.keys()).difference(v1.keys())
    joint = set(v1.keys()) & set(v2.keys())

    # deal with __migrator: ordering
    if "__migrator" in v2:
        print(v2)
        ordering = v2["__migrator"].get("ordering", {})
        print(ordering)
    else:
        ordering = {}

    # special keys
    if "__migrator" in right:
        right.remove("__migrator")

    # special keys in joint
    special_variants = {}
    if "pin_run_as_build" in joint:
        # For run_as_build we enforce the migrator's pin
        # TODO: should this just be a normal ordering merge, favoring more exact pins?
        joint.remove("pin_run_as_build")
        special_variants["pin_run_as_build"] = {
            **v1["pin_run_as_build"],
            **v2["pin_run_as_build"],
        }

    if "zip_keys" in joint:
        # zip_keys is a bit weird to join on as we don't have a particularly good way of identifying
        # a block.  Longer term having these be named blocks would make life WAY simpler
        # That does require changes to conda-build itself though
        #
        # A zip_keys block is deemend mergeable if zkₛ,ᵢ ⊂ zkₘ,ᵢ
        zk_out = []
        zk_l = {frozenset(e) for e in v1["zip_keys"]}
        zk_r = {frozenset(e) for e in v2["zip_keys"]}

        for zk_r_i in sorted(zk_r, key=lambda x: -len(x)):
            for zk_l_i in sorted(zk_l, key=lambda x: -len(x)):
                # Merge the longest common zk first
                if zk_l_i.issubset(zk_r_i):
                    zk_l.remove(zk_l_i)
                    zk_r.remove(zk_r_i)
                    zk_out.append(zk_r_i)
                    break
            else:
                # Nothing to do
                pass

        zk_out.extend(zk_l)
        zk_out.extend(zk_r)
        zk_out = sorted(
            [sorted(zk) for zk in zk_out], key=lambda x: (len(x), str(x))
        )

        joint.remove("zip_keys")
        special_variants["zip_keys"] = zk_out

    joint_variant = {}
    for k in joint:
        v_left, v_right = ensure_list(v1[k]), ensure_list(v2[k])
        joint_variant[k] = variant_key_add(
            k, v_left, v_right, ordering=ordering.get(k, None)
        )

    out = {
        **toolz.keyfilter(lambda k: k in left, v1),
        **toolz.keyfilter(lambda k: k in right, v2),
        **joint_variant,
        **special_variants,
    }

    return out
