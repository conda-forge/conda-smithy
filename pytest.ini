[pytest]
markers =
    legacy_circle: Test designed to run as if prior to the azure migration
    legacy_travis: Test designed to run as if prior to the azure migration
    legacy_appveyor: Test designed to run as if prior to the azure migration
    cli: CLI tests outside of test/test_cli.py