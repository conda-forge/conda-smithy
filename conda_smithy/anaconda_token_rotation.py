"""This module updates/rotates anaconda/binstar tokens.

The correct way to use this module is to call its functions via the command
line utility. The relevant one is

    conda-smithy update-anaconda-token

Note that if you are using appveyor, you will need to push the changes to the
conda-forge.yml in your feedstock to GitHub.
"""

import os
import sys
from contextlib import redirect_stderr, redirect_stdout
from typing import Optional, Union

import requests
from github import Github

from conda_smithy.utils import update_conda_forge_config


def _get_anaconda_token():
    """use this helper to enable easier patching for tests"""
    try:
        from conda_smithy.ci_register import anaconda_token

        return anaconda_token
    except ImportError:
        raise RuntimeError(
            "You must have the anaconda token defined to do token rotation!"
        )


def rotate_anaconda_token(
    user: str,
    project: str,
    feedstock_config_path: Optional[str],
    drone: bool = True,
    circle: bool = True,
    travis: bool = True,
    azure: bool = True,
    appveyor: bool = True,
    github_actions: bool = True,
    token_name: str = "BINSTAR_TOKEN",
    drone_endpoints: Union[list, tuple] = (),
):
    """Rotate the anaconda (binstar) token used by the CI providers

    All exceptions are swallowed and stdout/stderr from this function is
    redirected to `/dev/null`. Sanitized error messages are
    displayed at the end.

    If you need to debug this function, define `DEBUG_ANACONDA_TOKENS` in
    your environment before calling this function.
    """
    # we are swallong all of the logs below, so we do a test import here
    # to generate the proper errors for missing tokens
    # note that these imports cover all providers
    from .ci_register import travis_endpoint  # noqa
    from .azure_ci_utils import default_config  # noqa
    from conda_smithy.github import gh_token

    anaconda_token = _get_anaconda_token()

    if github_actions:
        gh = Github(gh_token())

    # capture stdout, stderr and suppress all exceptions so we don't
    # spill tokens
    failed = False
    err_msg = None
    with open(os.devnull, "w") as fp:
        if "DEBUG_ANACONDA_TOKENS" in os.environ:
            fpo = sys.stdout
            fpe = sys.stderr
        else:
            fpo = fp
            fpe = fp

        with redirect_stdout(fpo), redirect_stderr(fpe):
            try:
                if circle:
                    try:
                        rotate_token_in_circle(
                            user, project, anaconda_token, token_name
                        )
                    except Exception as e:
                        if "DEBUG_ANACONDA_TOKENS" in os.environ:
                            raise e
                        else:
                            err_msg = (
                                f"Failed to rotate token for {user}/{project}"
                                " on circle!"
                            )
                            failed = True
                            raise RuntimeError(err_msg)

                if drone:
                    for drone_endpoint in drone_endpoints:
                        try:
                            rotate_token_in_drone(
                                user,
                                project,
                                anaconda_token,
                                token_name,
                                drone_endpoint,
                            )
                        except Exception as e:
                            if "DEBUG_ANACONDA_TOKENS" in os.environ:
                                raise e
                            else:
                                err_msg = (
                                    f"Failed to rotate token for {user}/{project}"
                                    f" on drone endpoint {drone_endpoint}!"
                                )
                                failed = True
                                raise RuntimeError(err_msg)

                if travis:
                    try:
                        rotate_token_in_travis(
                            user,
                            project,
                            feedstock_config_path,
                            anaconda_token,
                            token_name,
                        )
                    except Exception as e:
                        if "DEBUG_ANACONDA_TOKENS" in os.environ:
                            raise e
                        else:
                            err_msg = (
                                f"Failed to rotate token for {user}/{project}"
                                " on travis!"
                            )
                            failed = True
                            raise RuntimeError(err_msg)

                if azure:
                    try:
                        rotate_token_in_azure(
                            user, project, anaconda_token, token_name
                        )
                    except Exception as e:
                        if "DEBUG_ANACONDA_TOKENS" in os.environ:
                            raise e
                        else:
                            err_msg = (
                                f"Failed to rotate token for {user}/{project}"
                                " on azure!"
                            )
                            failed = True
                            raise RuntimeError(err_msg)

                if appveyor:
                    try:
                        rotate_token_in_appveyor(
                            feedstock_config_path, anaconda_token, token_name
                        )
                    except Exception as e:
                        if "DEBUG_ANACONDA_TOKENS" in os.environ:
                            raise e
                        else:
                            err_msg = (
                                f"Failed to rotate token for {user}/{project}"
                                " on appveyor!"
                            )
                            failed = True
                            raise RuntimeError(err_msg)

                if github_actions:
                    try:
                        rotate_token_in_github_actions(
                            user, project, anaconda_token, token_name, gh
                        )
                    except Exception as e:
                        if "DEBUG_ANACONDA_TOKENS" in os.environ:
                            raise e
                        else:
                            err_msg = (
                                f"Failed to rotate token for {user}/{project}"
                                " on github actions!"
                            )
                            failed = True
                            raise RuntimeError(err_msg)

            except Exception as e:
                if "DEBUG_ANACONDA_TOKENS" in os.environ:
                    raise e
                failed = True

    if failed:
        if err_msg:
            raise RuntimeError(err_msg)
        else:
            raise RuntimeError(
                f"Rotating the feedstock token in providers for {user}/{project} failed!"
                " Try the command locally with DEBUG_ANACONDA_TOKENS"
                " defined in the environment to investigate!"
            )


def rotate_token_in_circle(user, project, binstar_token, token_name):
    from conda_smithy.ci_register import circle_token

    url_template = (
        "https://circleci.com/api/v1.1/project/github/{user}/{project}/envvar{extra}?"
        "circle-token={token}"
    )

    r = requests.get(
        url_template.format(
            token=circle_token, user=user, project=project, extra=""
        )
    )
    if r.status_code != 200:
        r.raise_for_status()

    have_binstar_token = False
    for evar in r.json():
        if evar["name"] == token_name:
            have_binstar_token = True

    if have_binstar_token:
        r = requests.delete(
            url_template.format(
                token=circle_token,
                user=user,
                project=project,
                extra=f"/{token_name}",
            )
        )
        if r.status_code != 200:
            r.raise_for_status()

    data = {"name": token_name, "value": binstar_token}
    response = requests.post(
        url_template.format(
            token=circle_token, user=user, project=project, extra=""
        ),
        data,
    )
    if response.status_code != 201:
        raise ValueError(response)


def rotate_token_in_drone(
    user, project, binstar_token, token_name, drone_endpoint
):
    from conda_smithy.ci_register import drone_session

    session = drone_session(drone_endpoint)

    r = session.get(f"/api/repos/{user}/{project}/secrets")
    r.raise_for_status()
    have_binstar_token = False
    for secret in r.json():
        if token_name == secret["name"]:
            have_binstar_token = True

    if have_binstar_token:
        r = session.patch(
            f"/api/repos/{user}/{project}/secrets/{token_name}",
            json={"data": binstar_token, "pull_request": False},
        )
        r.raise_for_status()
    else:
        response = session.post(
            f"/api/repos/{user}/{project}/secrets",
            json={
                "name": token_name,
                "data": binstar_token,
                "pull_request": False,
            },
        )
        if response.status_code != 200:
            response.raise_for_status()


def rotate_token_in_travis(
    user, project, feedstock_config_path, binstar_token, token_name
):
    """update the binstar token in travis."""
    from conda_smithy.ci_register import (
        travis_endpoint,
        travis_get_repo_info,
        travis_headers,
    )

    headers = travis_headers()

    repo_info = travis_get_repo_info(user, project)
    repo_id = repo_info["id"]

    r = requests.get(
        f"{travis_endpoint}/repo/{repo_id}/env_vars",
        headers=headers,
    )
    if r.status_code != 200:
        r.raise_for_status()

    have_binstar_token = False
    ev_id = None
    for ev in r.json()["env_vars"]:
        if ev["name"] == token_name:
            have_binstar_token = True
            ev_id = ev["id"]

    data = {
        "env_var.name": token_name,
        "env_var.value": binstar_token,
        "env_var.public": "false",
    }

    if have_binstar_token:
        r = requests.patch(
            f"{travis_endpoint}/repo/{repo_id}/env_var/{ev_id}",
            headers=headers,
            json=data,
        )
        r.raise_for_status()
    else:
        r = requests.post(
            f"{travis_endpoint}/repo/{repo_id}/env_vars",
            headers=headers,
            json=data,
        )
        if r.status_code != 201:
            r.raise_for_status()

    # we remove the token in the conda-forge.yml since on travis the
    # encrypted values override any value we put in the API
    with update_conda_forge_config(feedstock_config_path) as code:
        if (
            "travis" in code
            and "secure" in code["travis"]
            and token_name in code["travis"]["secure"]
        ):
            del code["travis"]["secure"][token_name]

            if len(code["travis"]["secure"]) == 0:
                del code["travis"]["secure"]

            if len(code["travis"]) == 0:
                del code["travis"]

            print(
                "An old value of the variable %s for travis was found in the "
                "conda-forge.yml. You may need to rerender this feedstock to "
                "use the new value since encrypted secrets inserted in travis.yml "
                "files override those set in the UI/API!"
            )


def rotate_token_in_azure(user, project, binstar_token, token_name):
    from vsts.build.v4_1.models import BuildDefinitionVariable

    from conda_smithy.azure_ci_utils import (
        build_client,
        get_default_build_definition,
    )
    from conda_smithy.azure_ci_utils import default_config as config

    bclient = build_client()

    existing_definitions = bclient.get_definitions(
        project=config.project_name, name=project
    )
    if existing_definitions:
        assert len(existing_definitions) == 1
        ed = existing_definitions[0]
    else:
        raise RuntimeError(
            f"Cannot add {token_name} to a repo that is not already registerd on azure CI!"
        )

    ed = bclient.get_definition(ed.id, project=config.project_name)

    if not hasattr(ed, "variables") or ed.variables is None:
        variables = {}
    else:
        variables = ed.variables

    variables[token_name] = BuildDefinitionVariable(
        allow_override=False,
        is_secret=True,
        value=binstar_token,
    )

    build_definition = get_default_build_definition(
        user,
        project,
        config=config,
        variables=variables,
        id=ed.id,
        revision=ed.revision,
    )

    bclient.update_definition(
        definition=build_definition,
        definition_id=ed.id,
        project=ed.project.name,
    )


def rotate_token_in_appveyor(feedstock_config_path, binstar_token, token_name):
    from conda_smithy.ci_register import appveyor_token

    headers = {"Authorization": f"Bearer {appveyor_token}"}
    url = "https://ci.appveyor.com/api/account/encrypt"
    response = requests.post(
        url, headers=headers, data={"plainValue": binstar_token}
    )
    if response.status_code != 200:
        raise ValueError(response)

    with update_conda_forge_config(feedstock_config_path) as code:
        code.setdefault("appveyor", {}).setdefault("secure", {})[
            token_name
        ] = response.content.decode("utf-8")


def rotate_token_in_github_actions(
    user, project, binstar_token, token_name, gh
):
    repo = gh.get_repo(f"{user}/{project}")
    assert repo.create_secret(token_name, binstar_token)
