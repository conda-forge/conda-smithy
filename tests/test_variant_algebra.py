import pytest
from textwrap import dedent

from conda_smithy.variant_algebra import parse_variant, variant_add


tv1 = parse_variant(
    """\
foo:
    - 1.10
bar:
    - 2
"""
)

tv2 = parse_variant(
    """\
foo:
    - 1.2
bar:
    - 3
"""
)

tv3 = parse_variant(
    """\
baz:
    - 1
bar:
    - 3
"""
)

tv4 = parse_variant(
    """\
baz:
    - 1
bar:
    - 0
    - 6
"""
)


def test_add():

    variant_add(tv1, tv2)

    # %%

    variant_add(tv1, tv3)

    # %%

    variant_add(tv2, tv3)
    # %%

    variant_add(tv1, variant_add(tv2, tv3))

    # %%

    variant_add(tv1, tv4)

    # %%

    variant_add(tv4, tv1)


def test_ordering():
    start = parse_variant(
        dedent(
            """\
    c_compiler:
        - toolchain
    """
        )
    )

    mig_compiler = parse_variant(
        dedent(
            """\
    __migrator:
        ordering:
            c_compiler:
                - toolchain
                - gcc
    c_compiler:
        - gcc
    """
        )
    )

    res = variant_add(start, mig_compiler)
    assert res["c_compiler"] == ["gcc"]
    print(res)
    # raise Exception()


def test_no_ordering():
    start = parse_variant(
        dedent(
            """\
    xyz:
        - 1
    """
        )
    )

    mig_compiler = parse_variant(
        dedent(
            """\
    __migrator:
        kind:
            version
        migration_no:
            1
    xyz:
        - 2
    """
        )
    )

    res = variant_add(start, mig_compiler)
    assert res["xyz"] == ["2"]
    print(res)
    # raise Exception()


def test_ordering_downgrade():
    start = parse_variant(
        dedent(
            """\
    jpeg:
        - 3.0
    """
        )
    )

    mig_compiler = parse_variant(
        dedent(
            """\
    __migrator:
        ordering:
            jpeg:
                - 3.0
                - 2.0
    jpeg:
        - 2.0
    """
        )
    )

    res = variant_add(start, mig_compiler)
    assert res["jpeg"] == ["2.0"]
    print(res)


def test_new_pinned_package():
    start = parse_variant(
        dedent(
            """\
    pin_run_as_build:
        jpeg:
            max_pin: x
    jpeg:
        - 3.0
    """
        )
    )

    mig_compiler = parse_variant(
        dedent(
            """\
    pin_run_as_build:
        gprc-cpp:
            max_pin: x.x
    gprc-cpp:
        - 1.23
    """
        )
    )

    res = variant_add(start, mig_compiler)
    assert res["gprc-cpp"] == ["1.23"]
    assert res["pin_run_as_build"]["gprc-cpp"]["max_pin"] == "x.x"
    print(res)


def test_zip_keys():
    start = parse_variant(
        dedent(
            """\
    zip_keys:
        -
            - vc
            - python
        -
            - qt
            - pyqt
    """
        )
    )

    mig_compiler = parse_variant(
        dedent(
            """\
    zip_keys:
        -
            - python
            - vc
            - vc_runtime
        -
            - root
            - c_compiler
    """
        )
    )

    res = variant_add(start, mig_compiler)
    print(res)

    assert len(res["zip_keys"]) == 3
    assert ["python", "vc", "vc_runtime"] in res["zip_keys"]


def test_migrate_windows_compilers():
    start = parse_variant(
        dedent(
            """
        c_compiler:
            - vs2008
            - vs2015
        vc:
            - '9'
            - '14'
        zip_keys:
            - - vc
              - c_compiler
        """
        )
    )

    mig = parse_variant(
        dedent(
            """
        c_compiler:
            - vs2008
            - vs2017
        vc:
            - '9'
            - '14.1'
        """
        )
    )

    res = variant_add(start, mig)
    print(res)

    assert len(res["c_compiler"]) == 2
    assert res["c_compiler"] == ["vs2008", "vs2017"]
    assert len(res["zip_keys"][0]) == 2


def test_pin_run_as_build():
    start = parse_variant(
        dedent(
            """\
    pin_run_as_build:
        python:
            max_pin: x.x
        boost-cpp:
            max_pin: x
    """
        )
    )

    mig_compiler = parse_variant(
        dedent(
            """\
    pin_run_as_build:
        boost-cpp:
            max_pin: x.x
        rust:
            max_pin: x
    """
        )
    )

    res = variant_add(start, mig_compiler)
    print(res)

    assert len(res["pin_run_as_build"]) == 3
