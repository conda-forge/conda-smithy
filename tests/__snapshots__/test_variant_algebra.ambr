# serializer version: 1
# name: test_migrate_windows_compilers
  dict({
    'c_compiler': list([
      'vs2008',
      'vs2017',
    ]),
    'migrator_ts': list([
      -1.0,
    ]),
    'vc': list([
      '9',
      '14.1',
    ]),
    'zip_keys': list([
      list([
        'vc',
        'c_compiler',
      ]),
    ]),
  })
# ---
# name: test_multiple_key_add_migration[res2]
  dict({
    'migrator_ts': list([
      -1.0,
    ]),
    'numpy': list([
      '1.16',
      '1.100',
      '1.200',
      '1.16',
      '1.16',
      '1.18',
    ]),
    'python': list([
      '3.6.* *_cpython',
      '3.9.* *_cpython',
      '3.10.* *_cpython',
      '3.7.* *_cpython',
      '3.8.* *_cpython',
      '3.6.* *_73_pypy',
    ]),
    'python_impl': list([
      'cpython',
      'cpython',
      'cpython',
      'cpython',
      'cpython',
      'pypy',
    ]),
    'zip_keys': list([
      list([
        'cuda_compiler_version',
        'docker_image',
      ]),
      list([
        'numpy',
        'python',
        'python_impl',
      ]),
    ]),
  })
# ---
# name: test_multiple_key_add_migration[res3]
  dict({
    'migrator_ts': -1.0,
    'python': list([
      '3.6.* *_cpython',
      '3.9.* *_cpython',
      '3.10.* *_cpython',
      '3.7.* *_cpython',
      '3.8.* *_cpython',
    ]),
    'python_impl': list([
      'cpython',
    ]),
    'zip_keys': list([
      list([
        'python',
      ]),
      list([
        'cuda_compiler_version',
        'docker_image',
      ]),
    ]),
  })
# ---
# name: test_new_pinned_package
  dict({
    'gprc-cpp': list([
      '1.23',
    ]),
    'jpeg': list([
      '3.0',
    ]),
    'migrator_ts': list([
      -1.0,
    ]),
    'pin_run_as_build': dict({
      'gprc-cpp': dict({
        'max_pin': 'x.x',
      }),
      'jpeg': dict({
        'max_pin': 'x',
      }),
    }),
  })
# ---
# name: test_no_ordering
  dict({
    'migrator_ts': list([
      -1.0,
    ]),
    'xyz': list([
      '2',
    ]),
  })
# ---
# name: test_ordering
  dict({
    'c_compiler': list([
      'gcc',
    ]),
    'migrator_ts': list([
      -1.0,
    ]),
  })
# ---
# name: test_ordering_downgrade
  dict({
    'jpeg': list([
      '2.0',
    ]),
    'migrator_ts': list([
      -1.0,
    ]),
  })
# ---
# name: test_ordering_space
  dict({
    'migrator_ts': list([
      -1.0,
    ]),
    'python': list([
      '2.7 *_cpython',
    ]),
  })
# ---
# name: test_pin_run_as_build
  dict({
    'migrator_ts': list([
      -1.0,
    ]),
    'pin_run_as_build': dict({
      'boost-cpp': dict({
        'max_pin': 'x.x',
      }),
      'python': dict({
        'max_pin': 'x.x',
      }),
      'rust': dict({
        'max_pin': 'x',
      }),
    }),
  })
# ---
# name: test_py39_migration[res2]
  dict({
    'migrator_ts': list([
      -1.0,
    ]),
    'numpy': list([
      '1.16',
      '1.100',
      '1.16',
      '1.16',
      '1.18',
    ]),
    'python': list([
      '3.6.* *_cpython',
      '3.9.* *_cpython',
      '3.7.* *_cpython',
      '3.8.* *_cpython',
      '3.6.* *_73_pypy',
    ]),
    'python_impl': list([
      'cpython',
      'cpython',
      'cpython',
      'cpython',
      'pypy',
    ]),
    'zip_keys': list([
      list([
        'cuda_compiler_version',
        'docker_image',
      ]),
      list([
        'numpy',
        'python',
        'python_impl',
      ]),
    ]),
  })
# ---
# name: test_py39_migration[res3]
  dict({
    'migrator_ts': -1.0,
    'python': list([
      '3.6.* *_cpython',
      '3.9.* *_cpython',
      '3.7.* *_cpython',
      '3.8.* *_cpython',
    ]),
    'python_impl': list([
      'cpython',
    ]),
    'zip_keys': list([
      list([
        'python',
      ]),
      list([
        'cuda_compiler_version',
        'docker_image',
      ]),
    ]),
  })
# ---
# name: test_variant_key_remove
  dict({
    'migrator_ts': -1.0,
    'numpy': list([
      '1.18',
      '1.16',
    ]),
    'python': list([
      '3.6.* *_73_pypy',
      '3.8.* *_cpython',
    ]),
    'python_impl': list([
      'pypy',
      'cpython',
    ]),
    'zip_keys': list([
      list([
        'python',
        'numpy',
        'python_impl',
      ]),
    ]),
  })
# ---
# name: test_zip_keys
  dict({
    'migrator_ts': list([
      -1.0,
    ]),
    'zip_keys': list([
      list([
        'c_compiler',
        'root',
      ]),
      list([
        'pyqt',
        'qt',
      ]),
      list([
        'python',
        'vc',
        'vc_runtime',
      ]),
    ]),
  })
# ---
